### DISCUSSION TITLE
Switch MQTT enable/disable Dynamic Power limiter

### DISCUSSION THREAD

**Discussion by isabido** [hoylabs/OpenDTU-OnBattery#583] [Ideas]
I am using OpenDTU with this wonderful option (Dynamic Power limiter). I would like to be able to enable and disable via MQTT for some automations in HomeAssistant that I am trying to use.

It would be a great improvement!

Thank you for your fantastic development.

**Comment by schlimmchen** [comment]
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/MQTT#power-limiter-topics

Does that satisfy your need?

**Reply by isabido** [reply]
Thank you for such a prompt response, I was looking at these options and if I am interpreting it correctly, it does not solve the need.

I really see it as simpler, publishing an enable/disable Dynamic Power Limiter switch via mqtt

Something like that?

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/64b7c55b-e6c4-4560-ab07-7e3b47a001e5)

**Reply by schlimmchen** [reply]
I don't understand... There is a topic into which you publish a value "1" or "0". The topic is "powerlimiter/cmd/mode". If the value is "1", the DPL will be disabled (and the inverter will be shut down), as if you disabled the DPL from the OpenDTU-OnBattery WebUI. If the value is "0", the DPL will operate as setup using the OpenDTU-OnBattery WebUI.

What about this does not work for you?

Thanks for the screenshot, however, I don't know what it is imitating and how to get a toggle switch working there. Make the toggle switch write "0" if enabled, and make it write "1" if disabled. That should be what you want.

**Reply by Manos1966** [reply]
Thanks for the explanations @schlimmchen 
I added an extra comment in the Wiki "Dynamic Power Limiter" regarding this.
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Dynamic-Power-Limiter

**Reply by isabido** [reply]
If I set the value to 1, the DPL is deactivated but it also turns off the inverter! I do not need that.

I need with DPL or without DPL but working at 100%. Forgive my English, I don't know if I can explain myself.

**Comment by Manos1966** [comment]
> If I set the value to 1, the DPL is deactivated but it also turns off the inverter! I do not need that.
> 
> I need with DPL or without DPL but working at 100%. Forgive my English, I don't know if I can explain myself.

@isabido  read the link schlimchen posted and focus on the function when value is set to "2" 😉

**Reply by isabido** [reply]
Oops, forgive my bad English, I just re-read it again with Google Translate and now I see that what I need is...

Activate DPL in the web interface, and then change between 0 (activated according to web) and 2 (DPL deactivated)

Tomorrow when the sun rises we'll try it! I'm going to create a helper BUTTON in HOMEASSISTANT, let's see if I remember how to do it!

Thank you all for the help. In any case, it would be interesting to facilitate it in future versions so that it appears directly as an entity in homeassistant.

**Reply by isabido** [reply]
It's already done, and it seems to work.

But the truth is that this involves creating an assistant with a button and two automations for the on and off position.

I think it would be an interesting improvement to publish it as an Entity directly in MQTT.

Thank you all for the support.

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/9d0a945b-8356-43b8-a7d6-a8284c41acfa)

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/07b0552f-5ae3-4a18-94be-0b8c1982073f)


![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/7a3aca44-2af0-4d13-97dc-d8df9ee72a80)

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/db5ce260-770b-4816-a963-10dbd22383ff)

**Reply by Manos1966** [reply]
Great Job @isabido 

Also nice to see that it is worthwhile to have the Wiki in English language (instead of German) 👍

**Reply by isabido** [reply]
When you send the parameter to value 2, the device turns off, NOT only disables the DPL. It behaves as if you were sending a 1.

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/f10c2125-205f-4884-80e3-a8cb32e3b4dd)

**Comment by schlimmchen** [comment]
that suggest that you are not using a VE.Direct charge controller, so the OpenDTUonBattery does not know how much solar power is being produced. ist that correct?

**Reply by isabido** [reply]
I do not use a battery installation, I use an HMS-2000-4T, what I want to achieve is to use DPL to avoid returning to the net, and deactivate the DPL to prioritize the surplus to heat water when necessary.

Hence the need to be able to activate or deactivate the DPL to be able to heat the water with a surplus manager (FreeDS).

**Reply by schlimmchen** [reply]
In that case mode 2 (unconditional full solar passthrough) does not work for you. That mode sets the inverter limit to the amount of solar power being produces, as reported by the battery charge controller (a Victron VE.Direct device).

So you want to avoid feeding power into the grid and use the DPL for that, understandable. Why do you need to disable the DPL when heating water? As soon as your power consumption rises due to the water heater, the DPL will turn up the limit of the inverter to match the additional need.

[A side question: I am looking to upgrade to a HMS-2000-4T. Are you happy with it? Is the radio interface reliable (I am disappointed that my HM-1500 needs a lot of retransmits of packets until a request is answered...)?]

**Reply by isabido** [reply]
It's going to be difficult for me to explain with my bad English.

Really what I need is to use all the maximum production to heat water as a priority, if the DPL is active as the grid balance is close to 0w, it does not start the surplus system, because it obviously does not see surpluses as the inverter is limited.

At the moment I have only had the HMS for 3 days, but the communication is very stable. At the moment I have not detected any problems. I will inform you!

**Reply by schlimmchen** [reply]
interesting setup. not an easy problem. how costly is it to start the heater and then probe whether or not there is surplus solar power? switch it back off if the inverter does not output more power once the limit was adjusted.

you can also publish 1 into /powerlimiter/cmd/mode, which disables the DPL and the inverter. then (you will need to wait until the inverter was switched off) switch the inverter back on using topic [serial]/cmd/power and set its limit using [serial]/cmd/limit_nonpersistent_relative to 100. you will then see whether or not there is surplus solar power.

unfortunately the inverter does not know or at least does not tell you how much energy it **could** produce if it was not limited. so the only way to probe that is to actually lift the limit (or there is something in place I don't know about).

maybe you can assume that there is surplus power once your grid consumption reaches 0W? your grid consumption will always be positive until there is sufficient solar power. and once there is, the probability that there is even more power is high. maybe wait until the grid consumption is less than 30W for 15min straight?

**Reply by isabido** [reply]
Really what I want to do is use all the "Inverter Total Power" value available and through a PID on the ESP32 control a resistive load (thermoelectric resistance) with this regulator.

Regulator 25A
https://es.aliexpress.com/item/20000003997748.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp

0-3.3v to 4-20mA converter
https://es.aliexpress.com/item/1005002960048002.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp

What I wanted to do is... the water temperature is below 45º, then it disables the DPL. And the regulation PID starts. When it reaches 65-70º, disconnect the PID and enable the DPL.

In order to do this, I just need to have this web switch visible through MQTT.
![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/559650/d822b512-7d9e-4978-a65b-7ca3c0308a05)

**Reply by spcqike** [reply]
@isabido what would happen, if you use like -20W as desired grid consumption? would your water heater start heating? or does it need a certain treshold?

**Reply by isabido** [reply]
As I have mentioned, I want to use all the solar power to heat water, and for that I have to use the "Inverter Total Power" variable in the PID. Apart from this resistance, if it does not have a minimum of 1000W, there is almost no interest in starting it. In that case, it would be preferable to use the surplus production for the rest of the household's consumption.

**Reply by schlimmchen** [reply]
And will the PID be part of the same ESP32 that runs (your custom version of) OpenDTU-OnBattery, or will it be another ESP32? If it is the same, you should of course hack it until you get what you want, i.e., disable the DPL. Maybe even hack it such that it does NOT stop the inverter when doing so. Otherwise... Hm, I think it would be acceptable (and simple) to implement a mode "3" of the DPL, which does what you want: "Free running inverter: Enable the inverter and set its limit to 100%". That's a special kind of "DPL is off" override. Would that be ideal for your situation?

**Reply by isabido** [reply]
The ideal would be to implement it in the same ESP where the OpenDTU is running. But I don't know if I will have the knowledge to be able to do it.

It would be easier for me to do it initially on another ESP32.

It would be very interesting to have this 3 mode! Thanks for your offer.

**Reply by Manos1966** [reply]
Hi @isabido 

> Really what I need is to use all the maximum production to heat water as a priority, if the DPL is active as the grid balance is close to 0w, it does not start the surplus system, because it obviously does not see surpluses as the inverter is limited.

Why do you use DPL? 
1. Do not use DPL
2. Send all produced electricity to the grid electricity-meter
3. You measure the electricity meter. It will show negative values. For example -400W
4. You know you are producing 400W more than what your house needs
5. Start the thermoelectric resistance with 400W
6. You electricity meter will show 0W

Some tipps: It is better to have many thermoelectric resistances of different size for example, one 250W, one 500W in order to be able to switch them on/off.
I know that in Greece you can order them and they are made to measure, they cost only 40EUR (I have two of them...)

The alternative is to use variable thermoelectric resistances that you can control per Software. 
Below is one I use in a test System and it works exceptionaly well. Unfortunately they are very expensive (compared to the 40EUR I pay to have them made in Greece):

> https://www.ta.co.at/en/x2-energy-management/immersion-heater-3000-w-variable-control
The EHS-R immersion heater (included with the ATON) can be controlled directly by the freely programmable controllers (UVR16x2, UVR610, ...) via PWM, with a variable output from 50 W to 3000 W. The immersion heater sends the sensor values back to the CAN-EZ3 for further utilisation or forwarding to the CAN bus or DL bus.

https://www.ta.co.at/aton/
The website has English

https://www.youtube.com/@TechnischeAlternativeRT
Video is only in German

**Reply by spcqike** [reply]
@Manos1966 he uses this 25A regulator, so it should be possible to limit the output power, shouldn't it?
https://es.aliexpress.com/item/20000003997748.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp

shouldn't it be possible, to just set the DPL to like -20W and let the regulator increase its output, whenever the grid consumption is < 0W? it would be slow, yeah, as it would need some DPL (and external regulator) cycles, but at the end it should open the DPL and consume the excess solar production. 

like grid -20W -> heat 20W -> grid 0 -> DPL increases to grid -20W -> heat 40W -> grid 0W ->  dpl increase ..... 

at least thats how i did my smart parkside charger, to charge my 20V 8Ah batterys. (with mqtt and node red)

> In that case, it would be preferable to use the surplus production for the rest of the household's consumption.

what other consumption do you have, if you already have a surplus? 

> if it does not have a minimum of 1000W, there is almost no interest in starting it

that kind of true. 20W heating isn't much for a big water tank. but it's also better than nothing. and, if it is possible to start the automatic heating with a desired grid consumption of like -20W or -50W, why shouldn't one try it? if it works, like it does in my case, you will end up with an "open" limiter. as your load (heater) increases with every calculation cycle.

**Reply by isabido** [reply]
Thanks for the options and recommendations, but I really don't want to use the GRID value.

I have a significant limitation with the language to make myself understood. Let's see with a practical case.

Let's imagine that I have a consumption of 1Kw at home, I have a solar production of 1Kw.

If I don't start the water heater, the balance is 0kw.

But I want to use all the solar energy produced to heat some water.

So if I turn on the water heater I would have 1Kw in the water heater, 1kw of consumption at home, and therefore a network consumption of 1Kw.

In short, my priority with the surplus is not to feed the house's consumption, it is to HEAT water, even if this implies consuming energy from the electrical grid.

With the options that you are considering, I would only heat the water with the excess energy after feeding all the consumers in the home. And I don't want to do that... the priority is to heat water and when finished, return to DPL mode to avoid draining the network.

**Reply by spcqike** [reply]
but, does the DPL matter in this scenario at all?

if you don't care about "only using excess energy", and if you don't mind to consume grid energy, why not just start the heater whenever needed?

the opendtu will see, that your grid consumption jumps from 0W to like 1000W and open up the DPL, which let it produce as much pv power as possible.

maybe it's just me, but where is the difference between your wish "disable DPL with a third mode, that just set it to 100%" and my "start your heater, the DPL will follow and likely end up at 100%" in your scenario?

**Reply by Manos1966** [reply]
@isabido I understand!

The  HMS-2000-4T will produce maximum 2000W electricity.
Is it a problem if all HMS-2000-4T electricity (2000W) goes to the GRID? 
Why do you want to use DPL? 

I have the same opinion as @spcqike 

1. Do not use DPL
2. Your  HMS-2000-4T produces as much electricity possible.
3. Your OpenDTU has the information on how much electricity the HMS-2000-4T produces
4. This information is also in Home Assistant
5. Use Home Assistant 
    - STOP DPL
    - Wait 2 Minutes (for the HMS-2000-4T to adjust to max available production)
          - IF water temperature is below 45º AND HMS-2000-4T produces > 400W THEN START thermoelectric resistance 
          - IF HMS-2000-4T produces < 400W THEN STOP thermoelectric resistance
          - IF water temperature is 65-70º THEN 
                 - STOP thermoelectric resistance 
                 - START DPL
     - Wait 10 minutes and then start again

**Comment by isabido** [comment]
No, I would like to do it like that...

First condition, do not send excesses to the GRID. (first because they don't pay me, and second that I only want to make the inverter work at the necessary power, to take care of the electronics of the inverter, in Spain the temperature in summer is very high)

If not, you will be able to add that 3 mode to enable/disable DPL.

I will set DPL to off always.

And now I will have to do some programming on an ESP32, which will do the following:

Temperature below 60º.
Start PID regulation adjusting based on "Inverter Total Power"

This way I avoid dumping into the grid, and I prioritize all the solar power to heat water.

Temperature above 60º
Start regulation PID based on "Grid Power"

This way it is not sent to the GRID, and when another consumer in the house demands energy it can be attended to.

In any case, I insist, I would like to have an entity in HA to enable/disable DPL.

**Reply by Manos1966** [reply]
Do I understand correctly that you can regulate the power of the 
https://es.aliexpress.com/item/20000003997748.html?spm=a2g0o.cart.0.0.eec37a9dSt4vWf&mp=1&gatewayAdapt=glo2esp
based on the 0-10V you send?

If yes, then you may want to have a look at the Shelly Plus Dimmer 0-10V
https://www.shelly.com/en-de/products/product-overview/shelly-plus-0-10-v-dimmer

It is much easier to control the 0-10V directly from Home Assistant without the need of an ESP32

Here is a video (in German 🙈 ) using the excess electricity to charge a battery 
https://www.youtube.com/watch?v=fcFFUN3Pkbo&t=4s

**Reply by isabido** [reply]
It is not a bad option to use this Shelly Dimmer, but the PID would have to run on HA or a NODERED.

The best thing would be for Shelly DIMMER himself to regulate the PID.

And the best better best!! It would be that OpenDTU integrated it, although you have to use the 0-3.3V to 0-10v or 4-20mA converter (I prefer the latter)

**Reply by Manos1966** [reply]
I would run everything on Home Assitant HA
- Data from OpenDTU regarding the HMS-2000-4T Inverter Total Power
- Shelly Plus Dimmer
- Water Temperasture
- Control of Shelly Plus Dimmer based on the conditions Inverter Total Power and Water Temperature

I do not think that OpenDTU can integrate the Water Temperature 😉 so this excludes OpenDTU

**Theoretically** you could run everything using SHELLY equipment, although I do not recommend this. 
HA is more powerful for such tasks.
Using only SHELLY equipment you need:
- Shelly Plug or Shelly Plus 1PM 
  connected to  HMS-2000-4T to measure electricity
- Shelly Plus 1PM with Sensors ro measure water Temperature
- Shelly Plus Dimmer to regulate the power

**Reply by isabido** [reply]
I am going to study your option of doing it in HA (with something like https://github.com/soloam/ha-pid-controller), although I would prefer to dedicate an ESP32 to perform the PID.

Regarding temperature control in opendtu, I see no problem integrating a DS18B20 type probe.

Even so, the PID will not control by temperature, it will do so following the power of the inverter and the power consumed by the thermos. That is, if the inverter generates 1,345kW, it will adjust so that the water heater consumes those 1,345kW.

The temperature will only be used to turn OFF the PID.

Even so, I would like to be able to have control of (enable/disable) DPL, since when the water reaches 70º, I will need to activate it to not send to the GRID.

I think now we have understood each other. Thank you all for the contributions and suggestions.

**Comment by isabido** [comment]
HI @schlimmchen Have you considered the option of adding this option in a future version. The famous MODE 3 that we were discussing to enable/disable DPL.

Thanks for your time!

**Reply by schlimmchen** [reply]
See #701.

**Reply by schlimmchen** [reply]
Hm, hold on, I have second thoughts about this... For two reasons: (1) The way this mode works will make people shoot themselves in the foot because it drains the battery unconditionally. I would need to adjust the implementation such that it considers the battery stop threshold. (2) I think there is a way to achieve what you want without a special DPL mode:

You are planning to manipulate the DPL using the mode command through MQTT. In that case, it should be sufficient to put the DPL in mode 1. This will disable the DPL. The DPL will shut the inverter down, and then leave it alone. You can then set the inverter limit to 100% (also through MQTT) and enable its power production.

When your water is warm, you can reset the DPL to mode 0 through MQTT (no need to change the inverter state).

What do you think, @isabido?
